### DISCUSSION TITLE
Configuration and Bootstrap for backend-ui-go

### DISCUSSION THREAD

**Discussion by mglmc** [edgexfoundry/community#246] [Q&A]
Hi,

we're working in a modified [backend-ui-go](https://github.com/edgexfoundry) and we realized that that module is a special module since it has a similar structure of the application service, but, without the advantages of using the bootstrap and configuration centralized (consul).

Any recommendation in how to add those advantages (consul and bootstrap) in the backend-ui-go?

**Comment by lenny-goodell** [comment]
@mglmc , the EdgeX UI uses bootstrapper, but not consul for it own configuration. It only uses the local configuration.yaml file due to `NOT` specifying the [-cp/--configProvider](https://docs.edgexfoundry.org/3.1/microservices/configuration/CommonCommandLineOptions/#config-provider) option in the docker file here:
https://github.com/edgexfoundry/edgex-ui-go/blob/main/Dockerfile#L57

Here is how all other services specify using Consul and to use default location of local configuration.yaml file on first startup which it uses to seed services configuration into Consul.
https://github.com/edgexfoundry/edgex-go/blob/main/cmd/core-data/Dockerfile#L56

Here is where the UI is using bootstrapper as all other EdgeX services do:
[bootstrapper](https://github.com/edgexfoundry/edgex-ui-go/blob/main/internal/service/service.go#L57)

Bootstrapper takes care of loading the configuration for either local configuration.yaml, URI to remote file or from Consul based on the command line options used.

**Reply by mglmc** [reply]
Thank you for the replay!

Sadly, the module by default doesn't seem to be prepared to create an instance of configuration.

We added: 

```yaml
ENTRYPOINT ["./edgex-ui-server","--configDir=res/docker", "-cp=consul.http://edgex-core-consul:8500", "--registry"]
```

in the Dockerfile and run it as usual with the rest of modules and the response was:

```txt
docker logs edgex-ui-go
panic: runtime error: index out of range [-1]

goroutine 1 [running]:
github.com/edgexfoundry/go-mod-bootstrap/v3/bootstrap/config.CreateProviderClient({0xd82608?, 0xc000496f90?}, {0xc00043c030?, 0x4?}, {0x0?, 0x4?}, 0x3?, {{0x7fff28e0ef24, 0x4}, {0x7fff28e0ef2b, ...}, ...})
        github.com/edgexfoundry/go-mod-bootstrap/v3@v3.2.0-dev.15/bootstrap/config/config.go:704 +0x2e6
github.com/edgexfoundry/go-mod-bootstrap/v3/bootstrap/config.(*Processor).loadCommonConfig(0xc0001bb130, {0x0, 0x0}, 0x0?, 0xc00049ca80, {0xd81bc8?, 0xc000194000}, {0xc58509, 0x5}, 0xcb5ef0)
        github.com/edgexfoundry/go-mod-bootstrap/v3@v3.2.0-dev.15/bootstrap/config/config.go:400 +0x18d
github.com/edgexfoundry/go-mod-bootstrap/v3/bootstrap/config.(*Processor).Process(0xc0001bb130, {0xc68384, 0x10}, {0xc58509, 0x5}, {0x0, 0x0}, {0xd81bc8, 0xc000194000}, {0x0, ...})
        github.com/edgexfoundry/go-mod-bootstrap/v3@v3.2.0-dev.15/bootstrap/config/config.go:169 +0x489
github.com/edgexfoundry/go-mod-bootstrap/v3/bootstrap.RunAndReturnWaitGroup({0xd7d7b0, 0xc00044e050}, 0xc000440530, {0xd821f8?, 0xc000530380}, {0xc68384, 0x10}, {0x0, 0x0}, {0xd81bc8, ...}, ...)
        github.com/edgexfoundry/go-mod-bootstrap/v3@v3.2.0-dev.15/bootstrap/bootstrap.go:126 +0x4af
github.com/edgexfoundry/go-mod-bootstrap/v3/bootstrap.Run({0xd7d7b0?, 0xc00044e050?}, 0xc000440530, {0xd821f8?, 0xc000530380?}, {0xc68384?, 0x53?}, {0x0?, 0xf?}, {0xd81bc8, ...}, ...)
        github.com/edgexfoundry/go-mod-bootstrap/v3@v3.2.0-dev.15/bootstrap/bootstrap.go:215 +0x10c
github.com/edgexfoundry/edgex-ui-go/internal/service.Main({0xd7d7b0, 0xc00044e050}, 0x60?, 0xc000000000)
        github.com/edgexfoundry/edgex-ui-go/internal/service/service.go:57 +0x5b1
main.main()
        github.com/edgexfoundry/edgex-ui-go/cmd/edgex-ui-server/main.go:26 +0x85
panic: runtime error: index out of range [-1]
```

So, it seems that the module needs extra modifications to be able to put the configuration in Consul.

**Reply by lenny-goodell** [reply]
What are you passing in for your configStem? 
Seems to be empty since the code below if failing.

```
	if configStem[len(configStem)-1] != '/' {
		configStem = configStem + "/"
	}
```
Note that all services now pass in `common.ConfigStemAll`
https://github.com/edgexfoundry/go-mod-core-contracts/blob/main/common/constants.go#L311
